### PR TITLE
Use web-thumbnailer to retrieve thumbnails

### LINKED ISSUE #345: Refactor thumbnail functions
### Refactoring
- [ ] move the following functions to `application/Thumbnail.php`:
  - [ ] `computeThumbnail`
  - [ ] `genThumbnail`
  - [ ] `lazyThumbnail`
  - [ ] `resizeImage`
  - [ ] `thumbnail`
- [ ] add test coverage
- [ ] apply coding conventions
### Improvements
- [ ] factorize `genThumbail` (duplicated code)
- [ ] refactor calls to 'get_http_url'
  - [ ] handle `Exception` in calling code
  - [ ] update unitary tests
- [ ] parametrizable thumbnail size

### LINKED ISSUE #712: Option to block non-https thumbnails
My Shaarli is now under HTTPS, but some thumbnails are under HTTP and I think is not a good thing for security and/or privacy.

Why thumbnails are not stored in local?

### LINKED ISSUE #1056: More options for theming picwall page
Today, in template picwall.html, `{$value.thumbnail}` populates a link with the image in it. Would that be possible to also get a variable containing only the URL of the image in order to improve theming options?

### PR DESCRIPTION
**Pull request by ArthurHoaro**
> Disclaimer: 
>   * this a proposition. ping @virtualtam @nodiscc 
>   * it requires PHP 5.6 (see #599)

Hi guys! I've tried to refactor thumbnails, and then end up writing [web-thumbnailer](https://github.com/ArthurHoaro/web-thumbnailer). It goes a bit beyond Shaarli needs but I still think we could use it as a dependency to retrieve thumbnail - [see the readme for more info](https://github.com/ArthurHoaro/web-thumbnailer).

This PR shows how to integrate web-thumbnailer within Shaarli. So, what changes:
  * a lot more thumbnails are retrieved
  * use blazy on linklist to lazy load images
  * thumbnails can be disabled
  * thumbs size is now 120x120
  * thumbs are now cropped to fit the expected size
  * again, requires PHP 5.6

Fixes #345 #425 #487 #543 #588 #590 

PS: a few things still need to be fixed, like a release on the WT repo.

EDIT: Fixes #712 - all thumbnails are now cached, no more hotlinking.
Fixes #1056

### CODE REVIEW

**Comment by nodiscc** [comment]
I have just read the patch comments, had a quick look at the implementation in Shaarli, and at the readme in https://github.com/ArthurHoaro/web-thumbnailer, and I would very much welcome this change. Lacking time to do a full code review though :/

Thanks

**Comment by kalvn** [comment]
May I suggest to add a way for templates to retrieve thumbnails direct URL? Or an array with several useful parameters such as what [computeThumbnail() function](https://github.com/shaarli/Shaarli/blob/master/index.php#L1770) does.

**Comment by ArthurHoaro** [comment]
@kalvn What's your use case?

**Comment by kalvn** [comment]
@ArthurHoaro the idea would be to allow more flexibility in the way templates display the thumbnail (with background-image instead of  for example). The current way is a bit rigid.

**Comment by ArthurHoaro** [comment]
@kalvn Each links will have a `thumbnail` field containing the thumbnail URL, or false if there isn't any.

PR updated and ready for v0.10!

File: tpl/default/js/shaarli.js [hunk 1]
```diff
@@ -321,8 +321,8 @@ window.onload = function () {
     /**
      * bLazy trigger
      */
-    var picwall = document.getElementById('picwall_container');
-    if (picwall != null) {
+    var blazyElements = document.getElementsByClassName('.b-lazy');
```
Comment (nicolasdanelon) [inline tpl/default/js/shaarli.js]: be careful ..
https://caniuse.com/#feat=getelementsbyclassname :)

File: tests/ThumbnailerTest.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    public function setUp()
+    {
+        $conf = new ConfigManager('tests/utils/config/configJson');
+        $conf->set('thumbnails.width', self::WIDTH);
+        $conf->set('thumbnails.height', self::HEIGHT);
+        $conf->set('dev.debug', true);
+
+        $this->thumbnailer = new Thumbnailer($conf);
+        // cache files in the sandbox
+        \WebThumbnailer\Application\ConfigManager::addFile('tests/utils/config/wt.json');
```
Comment (nicolasdanelon) [inline tests/ThumbnailerTest.php]: maybe you should use a `use` keyword? like you do in `ConfigManager`

**Comment by nodiscc** [comment]
> Each links will have a thumbnail field containing the thumbnail URL, or false if there isn't any.

Very nice.

 - [ ] I think we should document the list of fields somewhere, perhaps as part of a https://github.com/shaarli/Shaarli/blob/master/doc/md/Directory-structure.md update (new `Datastore structure` section). I thought there already was info on this somewhere, but can't find it. As far as I can tell currently Shaarli handles the following fields

```
created (??)
description (string)
id (integer)
private (boolean)
shorturl (string, 6-char long)
tags (array of strings)
title (string)
updated (???)
url (string)
```

add `thumbnails` to that.

File: tpl/default/js/shaarli.js (hunk 1 above)
Comment (ArthurHoaro) [reply tpl/default/js/shaarli.js]: Good to know... I've used it multiple times in this file.

**Comment by ArthurHoaro** [comment]
@nodiscc Good idea, I'll add that. And dates are now stored as `DateTime` objects.

File: index.php [hunk 3]
```diff
@@ -768,22 +769,51 @@ function renderPage($conf, $pluginManager, $LINKSDB, $history, $sessionManager)
     // -------- Picture wall
     if ($targetPage == Router::$PAGE_PICWALL)
     {
+        if (! $conf->get('thumbnails.enabled')) {
+            header('Location: ?');
```
Comment (nodiscc) [inline index.php]: Maybe display a warning message to the user `Picture wall unavailable (thumbnails are disabled)`. Or even better, hide the `Picture wall` button from the toolbar.

File: index.php [hunk 4]
```diff
@@ -768,22 +769,51 @@ function renderPage($conf, $pluginManager, $LINKSDB, $history, $sessionManager)
...
+                $LINKSDB[$key] = $item;
+                $updateDB = true;
+                $cpt++;
+            }
+
+            if (isset($link['thumbnail']) && $link['thumbnail'] !== false) {
+                $linksToDisplay[] = $link; // Add to array.
+            }
+
+            // If we retrieved new thumbnails, we update the database every 20 links.
```
Comment (nodiscc) [inline index.php]: Why not update the database at every new thumbnail retrieval? Performance reason?

File: index.php (hunk 3 above)
Comment (ArthurHoaro) [reply index.php]: Good idea. I did both.

File: index.php (hunk 4 above)
Comment (ArthurHoaro) [reply index.php]: Yes. Retrieving hundreds/thousands of thumbnails already takes a *very* long time, I didn't want to call serialize+gzip on every thumbnail. However, every 20 links is totally arbitrary.

Comment (ArthurHoaro) [reply index.php]: But I just noticed that having 2 variables is unnecessary.

**Comment by ArthurHoaro** [comment]
Rebased with the new frontend dependencies management, and added the documentation page suggested by @nodiscc.

Ready!

File: application/Thumbnailer.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+     * Retrieve a thumbnail for given URL
+     *
+     * @param string $url where to look for a thumbnail.
+     *
+     * @return bool|string The thumbnail relative cache file path, or false if none has been found.
+     */
+    public function get($url)
+    {
+        try {
+            return $this->wt->thumbnail($url);
```
Comment (virtualtam) [inline application/Thumbnailer.php]: this raises the following error when running tests:

```
$ composer install
# [...]

$ make test
# [...]

1) Shaarli\ThumbnailerTest::testThumbnailValid
Error: Call to undefined function WebThumbnailer\Utils\imagecreatefromstring()

/home/virtualtam/dev/shaarli/shaarli/vendor/arthurhoaro/web-thumbnailer/src/WebThumbnailer/Utils/ImageUtils.php:135
/home/virtualtam/dev/shaarli/shaarli/vendor/arthurhoaro/web-thumbnailer/src/WebThumbnailer/Finder/DefaultFinder.php:54
/home/virtualtam/dev/shaarli/shaarli/vendor/arthurhoaro/web-thumbnailer/src/WebThumbnailer/Application/Thumbnailer.php:95
/home/virtualtam/dev/shaarli/shaarli/vendor/arthurhoaro/web-thumbnailer/src/WebThumbnailer/WebThumbnailer.php:105
/home/virtualtam/dev/shaarli/shaarli/application/Thumbnailer.php:53
/home/virtualtam/dev/shaarli/shaarli/tests/ThumbnailerTest.php:50
```

as well as an error from HTTPD logs:

```
[Sat Mar 31 17:50:01.755957 2018] [php7:error] [pid 506] [client 127.0.0.1:46560] PHP Fatal error:  Uncaught Error: Call to undefined function WebThumbnailer\Utils\imagecreatefromstring() in /home/virtualtam/dev/shaarli/shaarli/vendor/arthurhoaro/web-thumbnailer/src/WebThumbnailer/Utils/ImageUtils.php:135
Stack trace:
#0 /home/virtualtam/dev/shaarli/shaarli/vendor/arthurhoaro/web-thumbnailer/src/WebThumbnailer/Finder/DefaultFinder.php(54): WebThumbnailer\Utils\ImageUtils::isImageString('<?xml version="...')
#1 /home/virtualtam/dev/shaarli/shaarli/vendor/arthurhoaro/web-thumbnailer/src/WebThumbnailer/Application/Thumbnailer.php(95): WebThumbnailer\Finder\DefaultFinder->find()
#2 /home/virtualtam/dev/shaarli/shaarli/vendor/arthurhoaro/web-thumbnailer/src/WebThumbnailer/WebThumbnailer.php(105): WebThumbnailer\Application\Thumbnailer->getThumbnail()
#3 /home/virtualtam/dev/shaarli/shaarli/application/Thumbnailer.php(53): WebThumbnailer\WebThumbnailer->thumbnail('https://httpd.a...')
#4 /home/virtualtam/dev/shaarli/shaarli/index.php(720): Shaarli\Thumbnailer->get('https://httpd.a...')
#5 /home/virtualtam/dev/sha in /home/virtualtam/dev/shaarli/shaarli/vendor/arthurhoaro/web-thumbnailer/src/WebThumbnailer/Utils/ImageUtils.php on line 135
```

Comment (ArthurHoaro) [reply application/Thumbnailer.php]: I suppose you don't have `php-gd`. I'll update the doc and add a check to disable thumbnails and display a message if that's the case.

Comment (virtualtam) [reply application/Thumbnailer.php]: `php-gd` is installed and enabled in `php.ini`, so this might be due to something else; I'll set up a fresh environment to see if it can be reproduced

Comment (ArthurHoaro) [reply application/Thumbnailer.php]: I'm really not sure, because I can reproduce your issue by disabling `php-gd`. Can you check the value of `extension_loaded('gd')` just to be sure?

**Comment by virtualtam** [comment]
Currently testing with a datastore containing around 3000 shaares, of which around 1000 are eligible for thumbnailing, some early remarks:
- it takes ages to build the thumbnail cache
- on the Picwall pages, thumbnail URIs are incorrect, this might be an escaping issue (my URI is `http://localhost.localdomain/~virtualtam/shaarli/`, using Apache HTTPD's `userdir` module)
- thumbnails are stored as PNG, switching to JPG would improve network bandwidth usage
- thumbnails are stored in a folder-file structure according to the original image URI:
    - this might be prone to nasty injections, hashing the URI or using UUIDs would be safer
    - this might cause issues on Windows due to unsupported characters and long paths

Note that I don't know to which extent these are related to WebThumbnailer or its usage in Shaarli, I will take the time to read the code more carefully (so please discard any remark that is irrelevant).

**Comment by virtualtam** [comment]
Additionally, do you know of ways to profile PHP code execution? I'd like to investigate where CPU time is spent :)

Hints so far:
- https://stackoverflow.com/questions/27038356/how-to-generate-flame-graphs-with-php
- https://github.com/jokkedk/webgrind
- https://pecl.php.net/package/xhprof
- https://github.com/humanmade/query-monitor-flamegraph

File: application/Thumbnailer.php (hunk 5 above)
Comment (ArthurHoaro) [reply application/Thumbnailer.php]: > gd is part of the Server requirements doc

Sure, but while it's still optional for Shaarli, it's now required for thumbnails.

> What about adding a "Server configuration check" page in the Tools section?

That's a good idea, and it should be used during the installation process. However, it's complementary (and should be addressed in another PR) because Shaarli shouldn't fail with an unexpected 500 error.

**Comment by ArthurHoaro** [comment]
Thank you for the feedback. Most of your remarks are related to WT itself. I'll focus on that, as I didn't really touched it for a while (this PR has been open for a long time!).

> it takes ages to build the thumbnail cache

I can improve that using [this](https://github.com/shaarli/Shaarli/commit/d65342e304f92643ba922200953cfebc51e1e482) feature. But downloading thousands of web pages will take ages in any case.

> on the Picwall pages, thumbnail URIs are incorrect, this might be an escaping issue

In Shaarli, thumbnails URI are stored as relative path to the cache file. I may have forgotten to check Shaarli is installed under a subfolder...

> thumbnails are stored as PNG, switching to JPG would improve network bandwidth usage

You're right, I can easily change that.

> this might be prone to nasty injections, hashing the URI or using UUIDs would be safer

Since it relies on `parse_url()`, I think it's safe, but you're right. I'm not sure why I chose this way to begin with.

> Additionally, do you know of ways to profile PHP code execution? I'd like to investigate where CPU time is spent :)

I never did it, but as you saw you can make xdebug dumping profiling files which can be analyzed using another tool. However, the time consumption is not a mystery. The process is pretty straightforward, for most links (aside from specific domains such as imgur, and a few other) :

  1. Download the remote page (long)
  2. Extract the thumbnail URI using a regex
  3. Download the thumbnail (long)
  4. Resize it using `php-gd` (I'm not sure how quick it is, analysis could be interesting here)
  5. Dump it in a cache file

So basically for 3k links and 1k thumb, you have something like 4k external `GET` request with a full download.

**Comment by ArthurHoaro** [comment]
I think I've handled all previous points. Also see [WT changelog](https://github.com/ArthurHoaro/web-thumbnailer/blob/master/CHANGELOG.md). Make sure to delete cache files before testing it again.

I've run a few test and a single thumb retrieval attempt takes between 400ms and 800ms on my dev environment.

**Comment by virtualtam** [comment]
Tested the latest version :)

> on the Picwall pages, thumbnail URIs are incorrect, this might be an escaping issue (my URI is http://localhost.localdomain/~virtualtam/shaarli/, using Apache HTTPD's userdir module)

I'm still facing this issue with `userdirs`, the images' `data-src` attribute contains a truncated value:

```html
<img data-src="haarli/cache/thumb/14dd5266c70789bdc806364df4586335/fcab71f58d58e34c7253c0f4132776b99f1c57ed125901.jpg" class="b-lazy b-error" src="#" alt="thumbnail" width="125" height="90">
```

This doesn't happen when serving Shaarli from a subdirectory, e.g. `http://localhost/dev/shaarli/`

It indeed took quite a lot of time to rebuild the thumbnail cache for the aforementioned 3k links collection, as I reached the web server's timeout several times for the operation to complete. This might partially be due to the fact that all URLs have become eligible to having a thumbnail generated, maybe this could be configurable?

For instance, I mostly care about having a thumbnail when the URI points to a video (youtube, vimeo, etc.), an image service (imgur, 500px, instagram, etc.) or an actual image; some websites also try to provide proper thumbnails. In other case the thumb is pretty random and/or not that useful.

**Comment by ArthurHoaro** [comment]
Thank you for your feedback. I guess there is something going wrong somewhere with the `~`. I'll try to reproduce your issue. 

That's an interesting suggestion. Instead of enabling/disabling thumbnails, I could add a third option, something like `Only popular media` which would drastically reduce the number of requests. Note that I currently mostly rely on [OpenGraph](http://ogp.me/) standard to retrieve the thumbnail, which is used by most news website. We would have to maintain that list, but it seems doable.

Regarding the timeouts and long time retrieval, I've had another idea. I could make a `Retrieve all thumbnails` button, which would open a dedicated view which retrieve all thumbs one by one through AJAX requests. It would fix any timeout issue and give the user feedback about what's going on, instead of letting the request running for XX minutes.

Depending on how long it takes me to handle these points, we might have to push the milestone on this.

**Comment by virtualtam** [comment]
Thumbnailing is the major remaining item for the [0.10.0](https://github.com/shaarli/Shaarli/milestones/0.10.0) milestone; if you're okay with it, we could re-scope this PR and related issues for 0.10.1, and release 0.10.0 with the current repository state?

**Comment by ArthurHoaro** [comment]
I don't mind pushing the milestone a bit for this PR, however it's a major feature and it does contain BC for 3rd party themes and plugins. It feels a bit strange to me to release it under a minor milestone. IMHO we should either wait for this PR to be ready, or have a short 0.10.x version and release it under 0.11.0.

**Comment by virtualtam** [comment]
> a short 0.10.x version and release it under 0.11.0

Sounds good to me, it will be nice to have some room between merging this PR and cutting a new release. This way, we could merge the PR in its current state and address further improvements to the thumbnailing features in new issues/PRs.

**Comment by ArthurHoaro** [comment]
I think I'm good here. I've:

  * Fixed the issue with Apache userdir
  * Auto-generate a `.htaccess` file in `thumbs/` cache folder which grants direct access
  * Move the sync to a dedicated page through AJAX requests (the picwall no longer trigger retrieval)
  * Added a `common` mode which only retrieve thumbs from a list of websites
  * Rebased the branch

As usual, clear the cache folder to test it. Regarding the milestone, since we didn't released  `v0.10` yet, maybe we can merge it before (depending on the feedback review of course).

**Comment by virtualtam** [comment]
Neat, I'll test it to(day|morrow) :)

**Comment by virtualtam** [comment]
Currently testing :)

- thumbs are correctly displayed when serving in a user-specific path (`http://domain.tld/~user/shaarli`)
- the new option to define which thumbnails are fetched is convenient, maybe it should default to "Only common media hosts" to stay consistent with the current behaviour
- the thumbnail update page is a nice addition, that should appear more prominently:
    - as a button in the Tools page (I actually missed it in its current form :p )
    - as a button in the Picture Wall page if the picture cache is empty
- the Picture Wall looks nicer with square thumbnails
- the thumbnails seem to be fetched only when browsing the linklist; browsing the picture wall has no effect

The following is most likely due to removing the `thumbnails` section under `data/config.json.php`:
- on the linklist, the `thumbnail` div does not respect its parent container's geometry, see screenshot below:
thumb-size

Here is the resulting configuration section, I think it misses the width/height settings:
```json
    "thumbnails": {
        "mode": "common"
    }
```

**Comment by ArthurHoaro** [comment]
> the new option to define which thumbnails are fetched is convenient, maybe it should default to "Only common media hosts" to stay consistent with the current behaviour

I'm not sure. The goal behind this PR, beside having a cleaner solution, is to provide a more complete feature. I feel like this option is nice to have for people who want a lighter solution, but displaying thumbnails for any website which provides an OpenGraph image is a good default option.

I don't really know what users would prefer, and this can easily be changed though, so any other opinion is welcome here.

> the thumbnail update page is a nice addition, that should appear more prominently

It should display a big `pure-alert` warning when you run the update or change the setting.

> as a button in the Tools page (I actually missed it in its current form :p )
> as a button in the Picture Wall page if the picture cache is empty

Sure, I'll add those.

For the last part:

  * I'll check the size settings update. I believe I never reset them after the update. And you probably already ran the update during your previous tests.
  * I'll fix the CSS issue if the thumb is too high.

Thanks again for the quick test :+1:

File: index.php [hunk 6]
```diff
@@ -1009,6 +1014,14 @@ function renderPage($conf, $pluginManager, $LINKSDB, $history, $sessionManager,
             $conf->set('api.secret', escape($_POST['apiSecret']));
             $conf->set('translation.language', escape($_POST['language']));
 
+            $thumbnailsMode = extension_loaded('gd') ? $_POST['enableThumbnails'] : Thumbnailer::MODE_NONE;
+            if ($conf->get('thumbnails.enabled', Thumbnailer::MODE_NONE) !== Thumbnailer::MODE_NONE) {
+                $_SESSION['warnings'][] = t(
+                    'You have enabled or changed thumbnails mode. <a href="?do=thumbs_update">Please synchonize them</a>.'
```
Comment (nodiscc) [inline index.php]: s/synchonize/synchronize

File: inc/web-thumbnailer.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{
+  "settings": {
+    "default": {
+      "download_mode": "HOTLINK",
```
Comment (nodiscc) [inline inc/web-thumbnailer.json]: Why not set the [download mode](https://github.com/ArthurHoaro/web-thumbnailer#download-mode) to `DOWNLOAD` ?

**Comment by ArthurHoaro** [comment]
Bunch of improvements for thumbnails integration:

  - add a default thumb size value (125x90px)
  - improve private vertical bar visual, especially with thumbnails
  - translations
  - add a sync thumbs button in tool and empty picwall page
  - fixes WT download mode in JSON config

I've changed a bit the way the orange bar for private links is displayed, mostly because it was truncated with floating thumbnails and no description. I went for an easier solution, and IMO it looks better. Example:

**Comment by ArthurHoaro** [comment]
Thank you @virtualtam and everyone for your time and reviews on this looong PR. Let's merge it! :tada:

**Comment by nicolasdanelon** [comment]
hi, I have a problem with thumbnails.. 

my list of links and the detail of a link refers to `public_html/links/cache/thumb/14dd5266c70789bdc806364df4586335/3a9726225cd4ed259458afa90126bb1509eea4a01259010.jpg`

the link should be  `links/cache/thumb/14dd5266c70789bdc806364df4586335/3a9726225cd4ed259458afa90126bb1509eea4a01259010.jpg`

### my shaarli is updated like this
```
$ cd /home/nicolasm/public_html/links
$ git checkout master
$ git pull --prune
$ composer install --no-dev --prefer-dist
$ yarn install
$ yarn run build

nginx  1.12.1
php-fpm 5.6.36-1

$ node -v
v6.14.2
$ yarn -v
1.9.2
$ npm -v
3.10.10

```
